components/raftstore/src/store/worker/split_controller.rs

```diff
@@ -81,11 +81,15 @@ where
     F: Fn(&mut T) -> &mut Vec<KeyRange>,
 {
     let mut sampled_key_ranges = vec![];
+    // Retain the non-empty key ranges.
+    // `key_ranges_provider` may return an empty key ranges vector, which will cause
+    // the later sampling to fall into a dead loop. So we need to filter it out here.
+    key_ranges_providers
+        .retain_mut(|key_ranges_provider| !key_ranges_getter(key_ranges_provider).is_empty());
```

Comment:
So it seems the key_ranges_providers won't too big so we run the function `key_ranges_getter` directly on item? If yes, Could we store the result of `key_range_getter(key_ranges_provider)`  and use it directly on L122?

Or If the `key_ranges_providers` maybe very big, we could consider to control  `while` loop instead?